bridge/gitlab/export.go

```diff
@@ -1,27 +1,514 @@
+	return client, nil
+}
+
 // ExportAll export all event made by the current user to Gitlab
-func (ge *gitlabExporter) ExportAll(repo *cache.RepoCache, since time.Time) (<-chan core.ExportResult, error) {
-	return nil, nil
+func (ge *gitlabExporter) ExportAll(ctx context.Context, repo *cache.RepoCache, since time.Time) (<-chan core.ExportResult, error) {
+	out := make(chan core.ExportResult)
+
+	user, err := repo.GetUserIdentity()
+	if err != nil {
+		return nil, err
+	}
+
+	ge.identityToken[user.Id().String()] = ge.conf[keyToken]
+
+	// get repository node id
+	ge.repositoryID = ge.conf[keyProjectID]
+
+	go func() {
+		defer close(out)
+
+		allIdentitiesIds := make([]entity.Id, 0, len(ge.identityToken))
+		for id := range ge.identityToken {
+			allIdentitiesIds = append(allIdentitiesIds, entity.Id(id))
+		}
+
+		allBugsIds := repo.AllBugsIds()
+
+		for _, id := range allBugsIds {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+				b, err := repo.ResolveBug(id)
+				if err != nil {
+					out <- core.NewExportError(err, id)
+					return
+				}
+
+				snapshot := b.Snapshot()
+
+				// ignore issues created before since date
+				// TODO: compare the Lamport time instead of using the unix time
+				if snapshot.CreatedAt.Before(since) {
+					out <- core.NewExportNothing(b.Id(), "bug created before the since date")
+					continue
+				}
+
+				if snapshot.HasAnyActor(allIdentitiesIds...) {
+					// try to export the bug and it associated events
+					ge.exportBug(ctx, b, since, out)
+				} else {
+					out <- core.NewExportNothing(id, "not an actor")
+				}
+			}
+		}
+	}()
+
+	return out, nil
+}
+
+// exportBug publish bugs and related events
+func (ge *gitlabExporter) exportBug(ctx context.Context, b *cache.BugCache, since time.Time, out chan<- core.ExportResult) {
+	snapshot := b.Snapshot()
+
+	var err error
+	var bugGitlabID int
+	var bugGitlabIDString string
+	var bugCreationId string
+
+	// Special case:
+	// if a user try to export a bug that is not already exported to Gitlab (or imported
+	// from Gitlab) and we do not have the token of the bug author, there is nothing we can do.
+
+	// skip bug if origin is not allowed
+	origin, ok := snapshot.GetCreateMetadata(core.KeyOrigin)
+	if ok && origin != target {
+		out <- core.NewExportNothing(b.Id(), fmt.Sprintf("issue tagged with origin: %s", origin))
+		return
+	}
+
+	// first operation is always createOp
+	createOp := snapshot.Operations[0].(*bug.CreateOperation)
+	author := snapshot.Author
+
+	// get gitlab bug ID
+	gitlabID, ok := snapshot.GetCreateMetadata(keyGitlabId)
+	if ok {
+		projectID, ok := snapshot.GetCreateMetadata(keyGitlabProject)
+		if !ok {
+			err := fmt.Errorf("expected to find gitlab project id")
+			out <- core.NewExportError(err, b.Id())
+			return
+		}
+
+		if projectID != ge.conf[keyProjectID] {
+			out <- core.NewExportNothing(b.Id(), "skipping issue imported from another repository")
+			return
+		}
+
+		_, ok = snapshot.GetCreateMetadata(keyGitlabUrl)
```

Comment:
not exporting because the gitlab url is missing doesn't help

Reply:
done